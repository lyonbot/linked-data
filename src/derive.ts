import { PatchOp, Path } from './types';
import { TrieNode } from './Trie';
import { shallowClone, cloneDeep } from './utils';
import type { applyPatches } from './applyPatches';

/**
 * @public
 */
export interface DeriveReport {
  mutated?: TrieNode<any>;
  result: any;
}

/**
 * @public
 */
export interface DeriveOptions {
  /**
   * 在遇到 set 和 spliceArray 的时候，是否对数据做 cloneDeep？
   *
   * （默认为 true）
   */
  clone?: boolean;

  /**
   * 是否对修改后的东西 freeze
   *
   * （默认为 true）
   */
  freeze?: boolean;

  /** 完成 Patches 在返回前会触发的回调 */
  onFinish?: (report: DeriveReport) => void;
}

/**
 * following immutable philosophy, make a patched new result.
 *
 * if you want to directly patch the src object, try {@link applyPatches} method
 *
 * @see {@link applyPatches}
 * @public
 */
export function derive(src: any, patches: PatchOp[], options?: DeriveOptions) {
  const onFinish = options && options.onFinish;

  if (!Array.isArray(patches) || patches.length === 0) {
    if (typeof onFinish === 'function') onFinish({ result: src });
    return src;
  }

  const mutated = new TrieNode<any>();
  mutated.value = shallowClone(src);

  const freeze = !!(options?.freeze ?? true);
  const processValue: (x: any) => any = options?.clone ?? true ? val => cloneDeep(val, { freeze }) : x => x;

  /**
   * @param path 必须指向一个 array / object
   */
  const getMutateNodeAt = (path: Path) => {
    let head = mutated;
    for (let i = 0; i < path.length; i++) {
      const key = path[i];
      const nextHead = head.getNodeAt([key], true);

      if (typeof nextHead.value === 'undefined') {
        const newChildValue = shallowClone(head.value![key]);
        head.value[key] = newChildValue;
        nextHead.value = newChildValue;
      }
      head = nextHead;
    }

    return head;
  };

  for (let i = 0; i < patches.length; i++) {
    const patch = patches[i];
    const mutateNode = getMutateNodeAt(patch.path);

    switch (patch.op) {
      case 'delete': {
        const { key, parent: parentNode } = mutateNode;

        if (!parentNode) {
          // 对根节点操作
          throw new Error('Cannot apply patch: impossible to delete root node.');
        }

        if (Array.isArray(parentNode.value)) {
          // 对数组操作
          throw new Error('Cannot apply patch: Path points to an array element.');
        }

        parentNode.deleteChild(key);
        delete parentNode.value![key];
        break;
      }

      case 'set': {
        const { key, parent: parentNode } = mutateNode;
        const newValue = processValue(patch.value);

        mutateNode.clearChildren(); // 当前字段，所有的子节点都被洗掉
        mutateNode.value = newValue;

        // 如果不是对根节点做操作，那么要回写到父级
        if (parentNode) parentNode.value![key] = newValue;
        break;
      }

      case 'resortArray': {
        const oldArray = mutateNode.value as Array<any>;
        const newArray = new Array(patch.indexMap.length);

        const oldNodeChildren = mutateNode.children;
        const newNodeChildren = {} as Record<string, typeof mutateNode>;

        for (let ni = 0; ni < patch.indexMap.length; ni++) {
          const oi: number = patch.indexMap[ni];
          if (oi === -1) continue; // do not reuse

          newArray[ni] = oldArray[oi];

          const oldChildNode = oldNodeChildren?.[oi];
          if (oldChildNode) {
            oldChildNode.setKey(String(ni));
            newNodeChildren[ni] = oldChildNode;
          }
        }

        mutateNode.value = newArray;
        mutateNode.children = newNodeChildren;

        const { key, parent: parentNode } = mutateNode;
        if (parentNode) parentNode.value![key] = newArray;

        break;
      }
    }
  }

  const result = mutated.value;
  if (freeze) {
    mutated.forEachDFS(node => {
      // eslint-disable-next-line no-param-reassign
      node.value = Object.freeze(node.value);
    });
  }

  if (typeof onFinish === 'function') onFinish({ mutated, result });
  return result;
}
