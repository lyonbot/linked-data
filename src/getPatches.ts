import { isObject, forEach, cloneDeep } from './utils';
import type { AnyObject, PatchOp, Path } from './types';

// a placeholder
const $taken = Symbol('taken');

interface GetPatchesOptions {
  root: any;
  oldRoot: any;
  getOld: (newObj: AnyObject) => AnyObject | null | undefined;
  processPatchValues?: (newObj: AnyObject) => AnyObject;
}

export function getPatches({
  root, //
  oldRoot,
  getOld,
  processPatchValues = cloneDeep,
}: GetPatchesOptions) {
  const wrapper = { root };
  const compareQueue: [AnyObject, AnyObject | null | undefined, Path | null][] = [
    [wrapper, root === oldRoot ? wrapper : { root: oldRoot }, null],
  ];

  const patches: PatchOp[] = [];
  const revertPatches: PatchOp[] = [];
  const ignoredNvs = new Set<any>();

  while (compareQueue.length) {
    const [nv, ov, vPath] = compareQueue.shift()!;

    if (ignoredNvs.has(nv)) continue;

    // queue sub-items
    forEach(nv, (value, key) => {
      if (!isObject(value)) return;
      compareQueue.push([value, getOld(value), vPath ? [...vPath, key] : []]);
    });

    // this object is not changed? skip it.
    if (!ov || ov === nv) continue;

    // this object is a changed array.
    // (smart hint: `path` is always available for array)
    if (Array.isArray(nv)) {
      const subPatches: PatchOp[] = [];
      const subRevertPatches: PatchOp[] = [];
      const oldArr = (ov as any[]).slice();

      let guessDelta = 0;
      let identical = nv.length === oldArr.length;
      const indexMap = nv.map((value, index) => {
        const guessedOi = index + guessDelta;
        const oi = oldArr[guessedOi] === value ? guessedOi : oldArr.indexOf(value);

        if (identical && oi !== index) identical = false;

        if (oi === -1) {
          if (isObject(value)) ignoredNvs.add(value);
          subPatches.push({
            op: 'set',
            path: [...vPath!, index],
            value: processPatchValues(value),
          });
          return -1;
        }

        guessDelta = oi - index;
        oldArr[oi] = $taken;
        return oi;
      });

      if (!identical) {
        subPatches.unshift({
          op: 'resortArray',
          path: vPath!,
          indexMap,
        });

        const revertIndexMap = oldArr.map((value, index) => {
          const ans = indexMap.indexOf(index);
          if (ans === -1) {
            subRevertPatches.push({
              op: 'set',
              path: [...vPath!, index],
              value: processPatchValues(value),
            });
          }

          return ans;
        });
        subRevertPatches.push({
          op: 'resortArray',
          path: vPath!,
          indexMap: revertIndexMap,
        });
      }

      patches.push(...subPatches);
      revertPatches.push(...subRevertPatches);
    } else {
      // this is just a changed object.
      const oKeys = new Set(Object.keys(ov));

      Object.keys(nv).forEach(key => {
        const value = (nv as any)[key];
        if (isObject(value)) ignoredNvs.add(value);

        const hadKey = oKeys.has(key);
        const needSet = !hadKey || value !== (ov as any)[key];

        oKeys.delete(key);
        if (needSet) {
          const path = vPath ? [...vPath, key] : [];
          patches.push({ op: 'set', path, value: processPatchValues(value) });

          if (hadKey) revertPatches.push({ op: 'set', path, value: processPatchValues((ov as any)[key]) });
          else revertPatches.push({ op: 'delete', path });
        }
      });

      oKeys.forEach(key => {
        const path = vPath ? [...vPath, key] : [];
        patches.push({ op: 'delete', path });
        revertPatches.push({ op: 'set', path, value: processPatchValues((ov as any)[key]) });
      });
    }
  }

  return { patches, revertPatches: revertPatches.reverse() };
}
