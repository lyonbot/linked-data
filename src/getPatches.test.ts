import { derive } from './derive';
import { getPatches } from './getPatches';

describe('getPatches', () => {
  it('basically works', () => {
    const newest = {
      foo: [1, 2, 3],
      bar: { baz: 'baz', newest: 'newest' },
    };
    const oldest = {
      foo: [0, 1, 2],
      bar: { baz: 'old baz', deleted: 'I am removed' },
    };

    const { patches, revertPatches } = getPatches(
      {
        root: newest, oldRoot: newest, getOld:
          n => {
            if (n === newest.foo)
              return [0, 1, 2];
            if (n === newest.bar)
              return { baz: 'old baz', deleted: 'I am removed' };
          }
      },
    );

    expect(patches).toMatchSnapshot();
    expect(revertPatches).toMatchSnapshot();
    // console.log(patches);
    // console.log(revertPatches);

    expect(derive(oldest, patches)).toEqual(newest);
    expect(derive(newest, revertPatches)).toEqual(oldest);
  });

  it('whole replaced', () => {
    const newest = { foo: [1, 2, 3] };
    const oldest = { hey: 'yo' };

    const { patches, revertPatches } = getPatches(
      {
        root: newest, oldRoot: { hey: 'yo' }, getOld:
          n => {
            if (n === newest.foo)
              return [0, 1, 2];
          }
      },
    );

    expect(patches).toMatchSnapshot();
    expect(revertPatches).toMatchSnapshot();
    // console.log(JSON.stringify(patches, null, 2));
    // console.log(JSON.stringify(revertPatches, null, 2));

    expect(derive(oldest, patches)).toEqual(newest);
    expect(derive(newest, revertPatches)).toEqual(oldest);
  });

  it('resorted objects', () => {
    const newest = [{ x: 0 }, { x: 1 }, { x: 2 }, { x: 3 }];
    const oldest = [{ x: -3 }, { x: 2 }, { x: 1 }, { y: 555 }];

    const { patches, revertPatches } = getPatches(
      {
        root: newest, oldRoot: newest, getOld:
          n => {
            if (n === newest)
              return [newest[3], newest[2], newest[1], { y: 555 }];
            if (n === newest[0])
              return { x: 999 }; // meanless because newest[0] not exist in original array
            if (n === newest[3])
              return { x: -3 };
          }
      },
    );

    expect(patches).toMatchSnapshot();
    expect(revertPatches).toMatchSnapshot();
    // console.log(JSON.stringify(patches, null, 2));
    // console.log(JSON.stringify(revertPatches, null, 2));

    expect(derive(oldest, patches)).toEqual(newest);
    expect(derive(newest, revertPatches)).toEqual(oldest);
  });
});
