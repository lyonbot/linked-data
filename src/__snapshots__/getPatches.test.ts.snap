// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPatches basically works 1`] = `
Array [
  Object {
    "indexMap": Array [
      1,
      2,
      -1,
    ],
    "op": "resortArray",
    "path": Array [
      "foo",
    ],
  },
  Object {
    "op": "set",
    "path": Array [
      "foo",
      2,
    ],
    "value": 3,
  },
  Object {
    "op": "set",
    "path": Array [
      "bar",
      "baz",
    ],
    "value": "baz",
  },
  Object {
    "op": "set",
    "path": Array [
      "bar",
      "newest",
    ],
    "value": "newest",
  },
  Object {
    "op": "delete",
    "path": Array [
      "bar",
      "deleted",
    ],
  },
]
`;

exports[`getPatches basically works 2`] = `
Array [
  Object {
    "op": "set",
    "path": Array [
      "bar",
      "deleted",
    ],
    "value": "I am removed",
  },
  Object {
    "op": "delete",
    "path": Array [
      "bar",
      "newest",
    ],
  },
  Object {
    "op": "set",
    "path": Array [
      "bar",
      "baz",
    ],
    "value": "old baz",
  },
  Object {
    "indexMap": Array [
      -1,
      0,
      1,
    ],
    "op": "resortArray",
    "path": Array [
      "foo",
    ],
  },
  Object {
    "op": "set",
    "path": Array [
      "foo",
      0,
    ],
    "value": 0,
  },
]
`;

exports[`getPatches resorted objects 1`] = `
Array [
  Object {
    "indexMap": Array [
      -1,
      2,
      1,
      0,
    ],
    "op": "resortArray",
    "path": Array [],
  },
  Object {
    "op": "set",
    "path": Array [
      0,
    ],
    "value": Object {
      "x": 0,
    },
  },
  Object {
    "op": "set",
    "path": Array [
      3,
      "x",
    ],
    "value": 3,
  },
]
`;

exports[`getPatches resorted objects 2`] = `
Array [
  Object {
    "op": "set",
    "path": Array [
      3,
      "x",
    ],
    "value": -3,
  },
  Object {
    "indexMap": Array [
      3,
      2,
      1,
      -1,
    ],
    "op": "resortArray",
    "path": Array [],
  },
  Object {
    "op": "set",
    "path": Array [
      3,
    ],
    "value": Object {
      "y": 555,
    },
  },
]
`;

exports[`getPatches whole replaced 1`] = `
Array [
  Object {
    "op": "set",
    "path": Array [],
    "value": Object {
      "foo": Array [
        1,
        2,
        3,
      ],
    },
  },
]
`;

exports[`getPatches whole replaced 2`] = `
Array [
  Object {
    "op": "set",
    "path": Array [],
    "value": Object {
      "hey": "yo",
    },
  },
]
`;
